base/set.jl

```diff
@@ -476,7 +476,19 @@ allunique(::Union{AbstractSet,AbstractDict}) = true
 
 allunique(r::AbstractRange) = !iszero(step(r)) || length(r) <= 1
 
-allunique(A::StridedArray) = length(A) < 32 ? _indexed_allunique(A) : _hashed_allunique(A)
+function allunique(A::StridedArray)
+    if length(A) < 32
+        _indexed_allunique(A)
+    elseif OrderStyle(eltype(A)) === Ordered()
+        if issorted(A) || issorted(A, rev=true)
+            all(i -> A[i] != A[i+1], 1:length(A)-1)
```

Comment:
Seems silly to take two passes over the array instead of one here.  Also, this assumes that the array has efficient linear indexing, and it incorrectly uses `!=` instead of `!isequal`.  Probably better to do something like
```jl
a1, rest1 = Iterators.peel(A)
a2, rest = Iterators.peel(rest1)
if isless(a1, a2)
    for a in rest
        if isless(a2, a)
            a2 = a
        elseif isequal(a2, a)
            return false
        else
            return _hashed_allunique(A)
        end
    end
elseif isless(a1, a2)
    for a in rest
        if isless(a, a2)
            a2 = a
        elseif isequal(a2, a)
            return false
        else
            return _hashed_allunique(A)
        end
    end
else
    return false
end
return true
```

Reply:
It's also not clear to me why this method is restricted to `StridedArray`, as opposed to `AbstractArray`.

Reply:
> It's also not clear to me why this method is restricted to StridedArray, as opposed to AbstractArray.

This was the conservative choice in #43375. The two concerns where arrays without fast linear indexing, and those like `mappedarray` for which indexing might be expensive. 

https://github.com/JuliaLang/julia/pull/43375#discussion_r766822917

https://github.com/JuliaLang/julia/pull/43375#discussion_r765401717